Autodiff code generation fails when caching values

```julia
using Reactant, Enzyme

mutable struct StateReturn
    st::Any
end

function simple_ad_fail(x, stret)
    loss = sum(x)
    stret.st = x .+ 1
    return loss
end

x = rand(10)
x_ra = x |> Reactant.to_rarray

Enzyme.gradient(Reverse, simple_ad_fail, x, Const(StateReturn(nothing)))

@code_hlo Enzyme.gradient(Reverse, simple_ad_fail, x_ra, Const(StateReturn(nothing)))
```

```julia
ERROR: AssertionError: No path found for TracedRArray{Float64,1N}(((1,),), size=(10,))
Stacktrace:
  [1] get_argidx(x::Reactant.TracedRArray{Float64, 1})
    @ Reactant.TracedUtils /mnt/software/lux/Reactant.jl/src/TracedUtils.jl:341
  [2] overload_autodiff(::ReverseMode{…}, ::Const{…}, ::Type{…}, ::Duplicated{…}, ::Const{…})
    @ Reactant /mnt/software/lux/Reactant.jl/src/Interpreter.jl:306
  [3] autodiff(::ReverseMode{…}, ::Const{…}, ::Type{…}, ::Duplicated{…}, ::Const{…})
    @ Reactant /mnt/software/lux/Reactant.jl/src/Overlay.jl:32
  [4] autodiff
    @ /mnt/.julia/packages/Enzyme/ydGh2/src/Enzyme.jl:524 [inlined]
  [5] macro expansion
    @ /mnt/.julia/packages/Enzyme/ydGh2/src/sugar.jl:270 [inlined]
  [6] gradient
    @ /mnt/.julia/packages/Enzyme/ydGh2/src/sugar.jl:258 [inlined]
  [7] gradient(rm::ReverseMode{…}, f::typeof(simple_ad_fail), x::Reactant.TracedRArray{…}, args::Tuple{…})
    @ Reactant ./<missing>:0
  [8] GenericMemory
    @ ./boot.jl:516 [inlined]
  [9] IdDict
    @ ./iddict.jl:31 [inlined]
 [10] IdDict
    @ ./iddict.jl:49 [inlined]
 [11] make_zero (repeats 2 times)
    @ /mnt/.julia/packages/EnzymeCore/15Zff/src/EnzymeCore.jl:529 [inlined]
 [12] macro expansion
    @ /mnt/.julia/packages/Enzyme/ydGh2/src/sugar.jl:316 [inlined]
 [13] gradient
    @ /mnt/.julia/packages/Enzyme/ydGh2/src/sugar.jl:258 [inlined]
 [14] (::Reactant.TracedUtils.var"#8#18"{Bool, Bool, typeof(Enzyme.gradient), Tuple{…}, Vector{…}, Tuple{…}})()
    @ Reactant.TracedUtils /mnt/software/lux/Reactant.jl/src/TracedUtils.jl:210
 [15] block!(f::Reactant.TracedUtils.var"#8#18"{…}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR /mnt/software/lux/Reactant.jl/src/mlir/IR/Block.jl:201
 [16] make_mlir_fn(f::Function, args::Tuple{…}, kwargs::Tuple{}, name::String, concretein::Bool; toscalar::Bool, return_dialect::Symbol, no_args_in_result::Bool, construct_function_without_args::Bool, do_transpose::Bool)
    @ Reactant.TracedUtils /mnt/software/lux/Reactant.jl/src/TracedUtils.jl:197
 [17] make_mlir_fn
    @ /mnt/software/lux/Reactant.jl/src/TracedUtils.jl:117 [inlined]
 [18] #10
    @ /mnt/software/lux/Reactant.jl/src/Compiler.jl:304 [inlined]
 [19] block!(f::Reactant.Compiler.var"#10#15"{typeof(Enzyme.gradient), Tuple{…}}, blk::Reactant.MLIR.IR.Block)
    @ Reactant.MLIR.IR /mnt/software/lux/Reactant.jl/src/mlir/IR/Block.jl:201
 [20] #9
    @ /mnt/software/lux/Reactant.jl/src/Compiler.jl:303 [inlined]
 [21] mmodule!(f::Reactant.Compiler.var"#9#14"{…}, blk::Reactant.MLIR.IR.Module)
    @ Reactant.MLIR.IR /mnt/software/lux/Reactant.jl/src/mlir/IR/Module.jl:92
 [22] compile_mlir!(mod::Reactant.MLIR.IR.Module, f::Function, args::Tuple{…}; optimize::Bool)
    @ Reactant.Compiler /mnt/software/lux/Reactant.jl/src/Compiler.jl:300
 [23] compile_mlir!
    @ /mnt/software/lux/Reactant.jl/src/Compiler.jl:299 [inlined]
 [24] #6
    @ /mnt/software/lux/Reactant.jl/src/Compiler.jl:290 [inlined]
 [25] context!(f::Reactant.Compiler.var"#6#7"{@Kwargs{…}, typeof(Enzyme.gradient), Tuple{…}}, ctx::Reactant.MLIR.IR.Context)
    @ Reactant.MLIR.IR /mnt/software/lux/Reactant.jl/src/mlir/IR/Context.jl:76
 [26] compile_mlir(f::Function, args::Tuple{…}; kwargs::@Kwargs{…})
    @ Reactant.Compiler /mnt/software/lux/Reactant.jl/src/Compiler.jl:288
 [27] top-level scope
    @ /mnt/software/lux/Reactant.jl/src/Compiler.jl:521
Some type information was truncated. Use `show(err)` to see complete types.
```

--- a/src/Compiler.jl
+++ b/src/Compiler.jl
@@ -631,17 +631,45 @@ function codegen_unflatten!(
                 path = path[2:end]
                 result_stores[path] = concrete_res_name
                 continue
-            else
-                @assert path[1] == :resargs
+            elseif path[1] == :resargs
                 unflatcode = :(args[$(path[2])])
                 path = path[3:end]
+            else
+                @show "path[1] == $(path[1]) has been ignored..." # XXX: Validate this is correct
+                continue
             end
 
             # unroll path tree
-            for p in path
+            for p in path[1:(end - 1)]
                 unflatcode = :(traced_getfield($unflatcode, $(Meta.quot(p))))
             end
-            unflatcode = :($unflatcode.data = $concrete_res_name)
+            if length(path) > 0
+                final_val = gensym("final_val")
+                unflatcode = quote
+                    $final_val = traced_getfield($unflatcode, $(Meta.quot(path[end])))
+                    if $final_val isa TracedRArray
+                        setfield!(
+                            $unflatcode,
+                            $(Meta.quot(path[end])),
+                            ConcreteRArray{eltype($final_val),ndims($final_val)}(
+                                $concrete_res_name, size($final_val)
+                            ),
+                        )
+                    elseif $final_val isa TracedRNumber
+                        setfield!(
+                            $unflatcode,
+                            $(Meta.quot(path[end])),
+                            ConcreteRNumber{eltype($final_val)}($concrete_res_name),
+                        )
+                    else
+                        setfield!($final_val, :data, $concrete_res_name)
+                    end
+                end
+            else
+                unflatcode = quote
+                    $unflatcode.data = $concrete_res_name
+                end
+            end
 
             push!(unflatten_code, unflatcode)
         end

--- a/src/Compiler.jl
+++ b/src/Compiler.jl
@@ -11,6 +11,7 @@ import ..Reactant:
     TracedRArray,
     TracedRNumber,
     RArray,
+    RNumber,
     OrderedIdDict,
     make_tracer,
     TracedToConcrete,
@@ -18,15 +19,18 @@ import ..Reactant:
     TracedType
 
 @inline traced_getfield(@nospecialize(obj), field) = Base.getfield(obj, field)
-@inline traced_getfield(@nospecialize(obj::AbstractArray), field) =
-    Base.getindex(obj, field)
-@inline traced_getfield(@nospecialize(obj::RArray), field) = Base.getfield(obj, field)
+@inline function traced_getfield(@nospecialize(obj::AbstractArray{T}), field) where {T}
+    (isbitstype(T) || obj isa RArray) && return Base.getfield(obj, field)
+    return Base.getindex(obj, field)
+end
 
 @inline traced_setfield!(@nospecialize(obj), field, val) = Base.setfield!(obj, field, val)
-@inline traced_setfield!(@nospecialize(obj::AbstractArray), field, val) =
-    Base.setindex!(obj, val, field)
-@inline traced_setfield!(@nospecialize(obj::RArray), field, val) =
-    Base.setfield!(obj, field, val)
+@inline function traced_setfield!(
+    @nospecialize(obj::AbstractArray{T}), field, val
+) where {T}
+    (isbitstype(T) || obj isa RArray) && return Base.setfield!(obj, field, val)
+    return Base.setindex!(obj, val, field)
+end
 
 function create_result(tocopy::T, path, result_stores) where {T}
     if !isstructtype(typeof(tocopy))
@@ -673,9 +677,9 @@ function codegen_unflatten!(
                             $clocal = if haskey($cache_dict, $final_val)
                                 $cache_dict[$final_val]
                             else
-                                $cache_dict[$final_val] = ConcreteRNumber{
-                                    eltype($final_val)
-                                }($concrete_res_name)
+                                $cache_dict[$final_val] = ConcreteRNumber{eltype($final_val)}(
+                                    $concrete_res_name
+                                )
                                 $cache_dict[$final_val]
                             end
                             traced_setfield!($unflatcode, $(Meta.quot(path[end])), $clocal)

--- a/src/Compiler.jl
+++ b/src/Compiler.jl
@@ -10,6 +10,7 @@ import ..Reactant:
     ConcreteRNumber,
     TracedRArray,
     TracedRNumber,
+    RArray,
     OrderedIdDict,
     make_tracer,
     TracedToConcrete,
@@ -19,10 +20,13 @@ import ..Reactant:
 @inline traced_getfield(@nospecialize(obj), field) = Base.getfield(obj, field)
 @inline traced_getfield(@nospecialize(obj::AbstractArray), field) =
     Base.getindex(obj, field)
+@inline traced_getfield(@nospecialize(obj::RArray), field) = Base.getfield(obj, field)
 
-@inline traced_setfield!(@nospecialize(obj), field, val) = Base.setfield!(obj, val, field)
+@inline traced_setfield!(@nospecialize(obj), field, val) = Base.setfield!(obj, field, val)
 @inline traced_setfield!(@nospecialize(obj::AbstractArray), field, val) =
     Base.setindex!(obj, val, field)
+@inline traced_setfield!(@nospecialize(obj::RArray), field, val) =
+    Base.setfield!(obj, field, val)
 
 function create_result(tocopy::T, path, result_stores) where {T}
     if !isstructtype(typeof(tocopy))

--- a/src/Compiler.jl
+++ b/src/Compiler.jl
@@ -17,9 +17,12 @@ import ..Reactant:
     TracedType
 
 @inline traced_getfield(@nospecialize(obj), field) = Base.getfield(obj, field)
-@inline traced_getfield(
-    @nospecialize(obj::AbstractArray{<:Union{ConcreteRNumber,ConcreteRArray}}), field
-) = Base.getindex(obj, field)
+@inline traced_getfield(@nospecialize(obj::AbstractArray), field) =
+    Base.getindex(obj, field)
+
+@inline traced_setfield!(@nospecialize(obj), field, val) = Base.setfield!(obj, val, field)
+@inline traced_setfield!(@nospecialize(obj::AbstractArray), field, val) =
+    Base.setindex!(obj, val, field)
 
 function create_result(tocopy::T, path, result_stores) where {T}
     if !isstructtype(typeof(tocopy))
@@ -573,32 +576,32 @@ function codegen_flatten!(linear_args, result_stores)
         push!(flatten_code, :($usbuf = $flatcode.data))
         push!(flatten_code, :($sbuf = XLA.synced_buffer($usbuf)))
 
-        # TODO
-        respaths = ((p for p in arg.paths if p[1] != :args)...,)
+        # TODO: unused for the time being
+        # respaths = ((p for p in arg.paths if p[1] == :result || p[1] == :resargs)...,)
 
         # resarg = false
-        for respath in respaths
-            if respath[1] == :result
-                flatcode = :result
-                respath = respath[2:end]
-                result_stores[respath] = usbuf
-                resarg = true
-            else
-                @assert respath[1] == :resargs
-                if respath[2] != path[2]
-                    continue
-                end
-                # flatcode = :(args[$(respath[2])])
-                path = path[3:end]
-            end
-            # for p in path
-            #     flatcode = :(traced_getfield($flatcode, $(Meta.quot(p))))
-            # end
-            # resarg = true
-            # flatcode = :($flatcode.data = $usbuf)
-            # @show flatcode
-            # push!(flatten_code, res)
-        end
+        # for respath in respaths
+        #     if respath[1] == :result
+        #         flatcode = :result
+        #         respath = respath[2:end]
+        #         result_stores[respath] = usbuf
+        #         resarg = true
+        #     else
+        #         @assert respath[1] == :resargs
+        #         if respath[2] != path[2]
+        #             continue
+        #         end
+        #         # flatcode = :(args[$(respath[2])])
+        #         path = path[3:end]
+        #     end
+        #     # for p in path
+        #     #     flatcode = :(traced_getfield($flatcode, $(Meta.quot(p))))
+        #     # end
+        #     # resarg = true
+        #     # flatcode = :($flatcode.data = $usbuf)
+        #     # @show flatcode
+        #     # push!(flatten_code, res)
+        # end
         # if resarg
         #     push!(resarg_code, :($usbuf = $flatcode.data))
         # end
@@ -624,53 +627,49 @@ function codegen_unflatten!(
 
     # mutate the result stores to point to the correct concrete results
     for (concrete_res_name, result) in zip(concretized_res_names, linear_results)
-        paths = ((p for p in result.paths if p[1] != :args)...,)
+        paths = ((p for p in result.paths if p[1] == :result || p[1] == :resargs)...,)
         for path in paths
             if path[1] == :result
                 unflatcode = :result
                 path = path[2:end]
                 result_stores[path] = concrete_res_name
                 continue
-            elseif path[1] == :resargs
+            else
+                @assert path[1] == :resargs
                 unflatcode = :(args[$(path[2])])
                 path = path[3:end]
-            else
-                continue
-            end
 
-            # unroll path tree
-            for p in path[1:(end - 1)]
-                unflatcode = :(traced_getfield($unflatcode, $(Meta.quot(p))))
-            end
-            if length(path) > 0
-                final_val = gensym("final_val")
-                unflatcode = quote
-                    $final_val = traced_getfield($unflatcode, $(Meta.quot(path[end])))
-                    if $final_val isa TracedRArray
-                        setfield!(
-                            $unflatcode,
-                            $(Meta.quot(path[end])),
-                            ConcreteRArray{eltype($final_val),ndims($final_val)}(
-                                $concrete_res_name, size($final_val)
-                            ),
-                        )
-                    elseif $final_val isa TracedRNumber
-                        setfield!(
-                            $unflatcode,
-                            $(Meta.quot(path[end])),
-                            ConcreteRNumber{eltype($final_val)}($concrete_res_name),
-                        )
-                    else
-                        setfield!($final_val, :data, $concrete_res_name)
+                for p in path[1:(end - 1)]
+                    unflatcode = :(traced_getfield($unflatcode, $(Meta.quot(p))))
+                end
+
+                if length(path) > 0
+                    final_val = gensym("final_val")
+                    unflatcode = quote
+                        $final_val = traced_getfield($unflatcode, $(Meta.quot(path[end])))
+                        if $final_val isa TracedRArray
+                            traced_setfield!(
+                                $unflatcode,
+                                $(Meta.quot(path[end])),
+                                ConcreteRArray{eltype($final_val),ndims($final_val)}(
+                                    $concrete_res_name, size($final_val)
+                                ),
+                            )
+                        elseif $final_val isa TracedRNumber
+                            traced_setfield!(
+                                $unflatcode,
+                                $(Meta.quot(path[end])),
+                                ConcreteRNumber{eltype($final_val)}($concrete_res_name),
+                            )
+                        else
+                            traced_setfield!($final_val, :data, $concrete_res_name)
+                        end
                     end
+                else
+                    unflatcode = :($unflatcode.data = $concrete_res_name)
                 end
-            else
-                unflatcode = quote
-                    $unflatcode.data = $concrete_res_name
-                end
+                push!(unflatten_code, unflatcode)
             end
-
-            push!(unflatten_code, unflatcode)
         end
     end
 

--- a/src/Compiler.jl
+++ b/src/Compiler.jl
@@ -635,7 +635,6 @@ function codegen_unflatten!(
                 unflatcode = :(args[$(path[2])])
                 path = path[3:end]
             else
-                @show "path[1] == $(path[1]) has been ignored..." # XXX: Validate this is correct
                 continue
             end
 

--- a/src/Compiler.jl
+++ b/src/Compiler.jl
@@ -627,7 +627,12 @@ function codegen_unflatten!(
     concrete_result,
     result_stores,
 )
-    unflatten_code = Expr[]
+    cache_dict = gensym("cache_dict")
+    unflatten_code = Expr[:(
+        $cache_dict = $(IdDict{
+            Union{TracedRArray,TracedRNumber},Union{ConcreteRArray,ConcreteRNumber}
+        }())
+    ),]
 
     # mutate the result stores to point to the correct concrete results
     for (concrete_res_name, result) in zip(concretized_res_names, linear_results)
@@ -649,22 +654,31 @@ function codegen_unflatten!(
 
                 if length(path) > 0
                     final_val = gensym("final_val")
+                    clocal = gensym("clocal")
                     unflatcode = quote
                         $final_val = traced_getfield($unflatcode, $(Meta.quot(path[end])))
                         if $final_val isa TracedRArray
-                            traced_setfield!(
-                                $unflatcode,
-                                $(Meta.quot(path[end])),
-                                ConcreteRArray{eltype($final_val),ndims($final_val)}(
+                            $clocal = if haskey($cache_dict, $final_val)
+                                $cache_dict[$final_val]
+                            else
+                                $cache_dict[$final_val] = ConcreteRArray{
+                                    eltype($final_val),ndims($final_val)
+                                }(
                                     $concrete_res_name, size($final_val)
-                                ),
-                            )
+                                )
+                                $cache_dict[$final_val]
+                            end
+                            traced_setfield!($unflatcode, $(Meta.quot(path[end])), $clocal)
                         elseif $final_val isa TracedRNumber
-                            traced_setfield!(
-                                $unflatcode,
-                                $(Meta.quot(path[end])),
-                                ConcreteRNumber{eltype($final_val)}($concrete_res_name),
-                            )
+                            $clocal = if haskey($cache_dict, $final_val)
+                                $cache_dict[$final_val]
+                            else
+                                $cache_dict[$final_val] = ConcreteRNumber{
+                                    eltype($final_val)
+                                }($concrete_res_name)
+                                $cache_dict[$final_val]
+                            end
+                            traced_setfield!($unflatcode, $(Meta.quot(path[end])), $clocal)
                         else
                             traced_setfield!($final_val, :data, $concrete_res_name)
                         end

--- a/src/Interpreter.jl
+++ b/src/Interpreter.jl
@@ -235,7 +235,7 @@ function overload_autodiff(
     primf = f.val
     primargs = ((v.val for v in args)...,)
 
-    fnwrap, func2, traced_result, result, seen_args, ret, linear_args, in_tys, linear_results = Reactant.TracedUtils.make_mlir_fn(
+    fnwrap, func2, traced_result, result, seen_args, ret, linear_args, in_tys, linear_results = TracedUtils.make_mlir_fn(
         primf, primargs, (), string(f) * "_autodiff", false
     )
 
@@ -302,7 +302,7 @@ function overload_autodiff(
                 cst = MLIR.IR.result(MLIR.Dialects.stablehlo.constant(; value=attr), 1)
                 push!(ad_inputs, cst)
             end
-        else
+        elseif TracedUtils.has_argidx(a)
             idx, path = TracedUtils.get_argidx(a)
             if idx == 1 && fnwrap
                 act = act_from_type(f, reverse, true)
@@ -321,6 +321,12 @@ function overload_autodiff(
                     continue
                 end
                 TracedUtils.push_val!(ad_inputs, args[idx].dval, path[3:end])
+            end
+        else
+            act = act_from_type(Enzyme.Const, reverse, true)
+            push!(ret_activity, act)
+            if act != enzyme_out && act != enzyme_outnoneed
+                continue
             end
         end
     end
@@ -385,7 +391,7 @@ function overload_autodiff(
                 end
                 residx += 1
             end
-        else
+        elseif TracedUtils.has_argidx(a)
             idx, path = TracedUtils.get_argidx(a)
             if idx == 1 && fnwrap
                 TracedUtils.set!(
@@ -405,6 +411,9 @@ function overload_autodiff(
                 )
                 residx += 1
             end
+        else
+            TracedUtils.set!(a, (), TracedUtils.transpose_val(MLIR.IR.result(res, residx)))
+            residx += 1
         end
     end
 

--- a/src/TracedUtils.jl
+++ b/src/TracedUtils.jl
@@ -341,6 +341,18 @@ function get_argidx(x)
     throw(AssertionError("No path found for $x"))
 end
 
+function has_argidx(x)
+    for path in x.paths
+        if length(path) == 0
+            continue
+        end
+        if path[1] == :args
+            return true
+        end
+    end
+    return false
+end
+
 function set!(x, path, tostore; emptypath=false)
     for p in path
         x = Reactant.Compiler.traced_getfield(x, p)

--- a/src/Tracing.jl
+++ b/src/Tracing.jl
@@ -6,32 +6,32 @@
     TracedSetPath = 5
 end
 
-for T in (
-    DataType,
-    Module,
-    Nothing,
-    Symbol,
-    AbstractChar,
-    AbstractFloat,
-    Integer,
-    AbstractString,
-    RArray,
-    RNumber,
-)
-    @eval function traced_type(::Type{T}, seen, mode) where {T<:$T}
+for T in (DataType, Module, Nothing, Symbol, AbstractChar, AbstractString, RArray, RNumber)
+    @eval function traced_type(::Type{T}, seen, mode, track_numbers) where {T<:$T}
         return T
     end
 end
 
-function traced_type(::Type{C}, seen::ST, mode::Val{Mode}) where {T,C<:Complex{T},ST,Mode}
+function traced_type(::Type{T}, seen, mode::Val{Mode}, track_numbers) where {T<:Number,Mode}
+    if Mode == ArrayToConcrete && any(Base.Fix1(<:, T), track_numbers)
+        return ConcreteRNumber{T}
+    end
+    return T
+end
+
+function traced_type(
+    ::Type{C}, seen::ST, mode::Val{Mode}, track_numbers::TN
+) where {T,C<:Complex{T},ST,Mode,TN}
     if !(C isa UnionAll)
-        return Complex{traced_type(T, seen, mode)}
+        return Complex{traced_type(T, seen, mode, track_numbers)}
     else
-        return @invoke traced_type(C::Type{Any}, seen::ST, mode::Val{Mode})
+        return @invoke traced_type(
+            C::Type{Any}, seen::ST, mode::Val{Mode}, track_numbers::TN
+        )
     end
 end
 
-function traced_type(::Type{T}, seen, mode) where {T<:Function}
+function traced_type(::Type{T}, seen, mode, track_numbers) where {T<:Function}
     # functions are directly returned
     if sizeof(T) == 0
         return T
@@ -41,7 +41,7 @@ function traced_type(::Type{T}, seen, mode) where {T<:Function}
     N = fieldcount(T)
     changed = false
     traced_fieldtypes = ntuple(Val(N)) do i
-        next = traced_type(fieldtype(T, i), seen, mode)
+        next = traced_type(fieldtype(T, i), seen, mode, track_numbers)
         changed |= next != fieldtype(T, i)
         next
     end
@@ -57,31 +57,34 @@ end
 @inline is_concrete_tuple(x::T2) where {T2} =
     (x <: Tuple) && !(x === Tuple) && !(x isa UnionAll)
 
-function traced_type(::Type{T}, seen, mode) where {T<:Tuple}
+function traced_type(::Type{T}, seen, mode, track_numbers) where {T<:Tuple}
     if !Base.isconcretetype(T) || !is_concrete_tuple(T) || T isa UnionAll
         throw(AssertionError("Type $T is not concrete type or concrete tuple"))
     elseif is_concrete_tuple(T) && any(T2 isa Core.TypeofVararg for T2 in T.parameters)
         # Tuple{((T2 isa Core.TypeofVararg ? Any : T2) for T2 in T.parameters)...}
         throw(AssertionError("Type tuple of vararg $T is not supported"))
     end
-    TT = [traced_type(T.parameters[i], seen, mode) for i in 1:length(T.parameters)]
+    TT = [
+        traced_type(T.parameters[i], seen, mode, track_numbers) for
+        i in 1:length(T.parameters)
+    ]
     return Tuple{TT...}
 end
 
-function traced_type(::Type{T}, seen, mode) where {N,V,T<:NamedTuple{N,V}}
-    return NamedTuple{N,traced_type(V, seen, mode)}
+function traced_type(::Type{T}, seen, mode, track_numbers) where {N,V,T<:NamedTuple{N,V}}
+    return NamedTuple{N,traced_type(V, seen, mode, track_numbers)}
 end
 
-function traced_type(::Type{T}, seen, mode) where {K,V,T<:AbstractDict{K,V}}
+function traced_type(::Type{T}, seen, mode, track_numbers) where {K,V,T<:AbstractDict{K,V}}
     dictty = T.name.wrapper
-    return dictty{K,traced_type(V, seen, mode)}
+    return dictty{K,traced_type(V, seen, mode, track_numbers)}
 end
 
 @inline getmap(::Val{T}) where {T} = nothing
 @inline getmap(::Val{T}, a, b, args...) where {T} = getmap(Val(T), args...)
 @inline getmap(::Val{T}, ::Val{T}, ::Val{T2}, args...) where {T,T2} = T2
 
-function traced_type(::Type{T}, seen, mode) where {T}
+function traced_type(::Type{T}, seen, mode, track_numbers) where {T}
     if T === Any
         return T
     end
@@ -110,7 +113,10 @@ function traced_type(::Type{T}, seen, mode) where {T}
     end
 
     if T isa Union
-        return Union{traced_type(T.a, seen, mode),traced_type(T.b, seen, mode)}
+        return Union{
+            traced_type(T.a, seen, mode, track_numbers),
+            traced_type(T.b, seen, mode, track_numbers),
+        }
     end
 
     # if abstract it must be by reference
@@ -133,7 +139,7 @@ function traced_type(::Type{T}, seen, mode) where {T}
     subTys = Type[]
     for f in 1:fieldcount(T)
         subT = fieldtype(T, f)
-        subTT = traced_type(subT, seen2, mode)
+        subTT = traced_type(subT, seen2, mode, track_numbers)
         changed |= subT != subTT
         push!(subTys, subTT)
     end
@@ -145,7 +151,7 @@ function traced_type(::Type{T}, seen, mode) where {T}
     subParms = []
     for SST in T.parameters
         if SST isa Type
-            TrT = traced_type(SST, seen, mode)
+            TrT = traced_type(SST, seen, mode, track_numbers)
             push!(subParms, TrT)
         else
             push!(subParms, SST)
@@ -163,7 +169,7 @@ function traced_type(::Type{T}, seen, mode) where {T}
         for f in 1:fieldcount(T)
             subT = fieldtype(T, f)
             subT2 = fieldtype(TT2, f)
-            subTT = traced_type(subT, seen3, mode)
+            subTT = traced_type(subT, seen3, mode, track_numbers)
             if subT2 != subTT
                 legal = false
                 break
@@ -178,7 +184,9 @@ function traced_type(::Type{T}, seen, mode) where {T}
     throw(NoFieldMatchError(T, TT2))
 end
 
-function traced_type(::Type{<:ConcreteRNumber{T}}, seen, ::Val{mode}) where {T,mode}
+function traced_type(
+    ::Type{<:ConcreteRNumber{T}}, seen, ::Val{mode}, track_numbers
+) where {T,mode}
     if mode == ConcreteToTraced
         return TracedRNumber{T}
     elseif mode == TracedToConcrete
@@ -188,7 +196,9 @@ function traced_type(::Type{<:ConcreteRNumber{T}}, seen, ::Val{mode}) where {T,m
     end
 end
 
-function traced_type(::Type{T}, seen, ::Val{mode}) where {T<:ConcreteRArray,mode}
+function traced_type(
+    ::Type{T}, seen, ::Val{mode}, track_numbers
+) where {T<:ConcreteRArray,mode}
     if mode == ConcreteToTraced
         @inline base_typet(TV::TT) where {TT<:UnionAll} =
             UnionAll(TV.var, base_typet(TV.body))
@@ -201,7 +211,9 @@ function traced_type(::Type{T}, seen, ::Val{mode}) where {T<:ConcreteRArray,mode
     end
 end
 
-function traced_type(::Type{T}, seen::ST, ::Val{mode}) where {ST,T<:TracedType,mode}
+function traced_type(
+    ::Type{T}, seen::ST, ::Val{mode}, track_numbers
+) where {ST,T<:TracedType,mode}
     T <: MissingTracedValue && error("TODO")
     if mode == ConcreteToTraced
         throw("TracedRArray $T cannot be traced")
@@ -218,26 +230,28 @@ function traced_type(::Type{T}, seen::ST, ::Val{mode}) where {ST,T<:TracedType,m
     end
 end
 
-function traced_type(::Type{T}, seen, mode) where {T<:XLAArray}
+function traced_type(::Type{T}, seen, mode, track_numbers) where {T<:XLAArray}
     throw("XLA $T array cannot be traced")
 end
 
-function traced_type(::Type{A}, seen::ST, ::Val{mode}) where {T,N,A<:Array{T,N},ST,mode}
+function traced_type(
+    ::Type{A}, seen::ST, ::Val{mode}, track_numbers
+) where {T,N,A<:Array{T,N},ST,mode}
     if mode == ArrayToConcrete && T <: ReactantPrimitive
         return ConcreteRArray{T,N}
     else
-        return Array{traced_type(T, seen, Val(mode)),N}
+        return Array{traced_type(T, seen, Val(mode), track_numbers),N}
     end
 end
 
 for P in (Ptr, Core.LLVMPtr, Base.RefValue)
-    @eval function traced_type(::Type{P}, seen, mode) where {T,P<:$P{T}}
-        return $P{traced_type(T, seen, mode)}
+    @eval function traced_type(::Type{P}, seen, mode, track_numbers) where {T,P<:$P{T}}
+        return $P{traced_type(T, seen, mode, track_numbers)}
     end
 end
 
-function traced_type(::Type{Val{T}}, seen, mode) where {T}
-    if traced_type(typeof(T), seen, mode) == typeof(T)
+function traced_type(::Type{Val{T}}, seen, mode, track_numbers) where {T}
+    if traced_type(typeof(T), seen, mode, track_numbers) == typeof(T)
         return Val{T}
     end
     throw("Val type $(Val{T}) cannot be traced")
@@ -274,12 +288,13 @@ function make_tracer(
     mode;
     toscalar=false,
     tobatch=nothing,
+    track_numbers=(),
     kwargs...,
 ) where {RT}
     if haskey(seen, prev)
         return seen[prev]
     end
-    TT = traced_type(RT, (), Val(mode))
+    TT = traced_type(RT, (), Val(mode), track_numbers)
     @assert !Base.isabstracttype(RT)
     @assert Base.isconcretetype(RT)
     nf = fieldcount(RT)
@@ -295,7 +310,16 @@ function make_tracer(
         for i in 1:nf
             if isdefined(prev, i)
                 xi = Base.getfield(prev, i)
-                xi2 = make_tracer(seen, xi, append_path(path, i), mode; toscalar, tobatch)
+                xi2 = make_tracer(
+                    seen,
+                    xi,
+                    append_path(path, i),
+                    mode;
+                    toscalar,
+                    tobatch,
+                    track_numbers,
+                    kwargs...,
+                )
                 if xi !== xi2
                     changed = true
                 end
@@ -318,7 +342,16 @@ function make_tracer(
     for i in 1:nf
         if isdefined(prev, i)
             xi = Base.getfield(prev, i)
-            xi2 = make_tracer(seen, xi, append_path(path, i), mode; toscalar, tobatch)
+            xi2 = make_tracer(
+                seen,
+                xi,
+                append_path(path, i),
+                mode;
+                toscalar,
+                tobatch,
+                track_numbers,
+                kwargs...,
+            )
             if xi !== xi2
                 changed = true
             end
@@ -543,7 +576,7 @@ function make_tracer(
 end
 
 function make_tracer(
-    seen, @nospecialize(prev::RT), @nospecialize(path), mode; kwargs...
+    seen, @nospecialize(prev::RT), @nospecialize(path), mode; track_numbers=(), kwargs...
 ) where {RT<:Array}
     if haskey(seen, prev)
         return seen[prev]
@@ -551,14 +584,14 @@ function make_tracer(
     if mode == ArrayToConcrete && eltype(RT) <: ReactantPrimitive
         return seen[prev] = ConcreteRArray(prev)
     end
-    TT = traced_type(eltype(RT), (), Val(mode))
+    TT = traced_type(eltype(RT), (), Val(mode), track_numbers)
     newa = Array{TT,ndims(RT)}(undef, size(prev))
     seen[prev] = newa
     same = true
     for I in eachindex(prev)
         if isassigned(prev, I)
             pv = prev[I]
-            nv = make_tracer(seen, pv, append_path(path, I), mode; kwargs...)
+            nv = make_tracer(seen, pv, append_path(path, I), mode; track_numbers, kwargs...)
             if pv !== nv
                 same = false
             end
@@ -584,12 +617,22 @@ function make_tracer(
 end
 
 function make_tracer(
-    seen, @nospecialize(prev::NamedTuple{A,RT}), @nospecialize(path), mode; kwargs...
+    seen,
+    @nospecialize(prev::NamedTuple{A,RT}),
+    @nospecialize(path),
+    mode;
+    track_numbers=(),
+    kwargs...,
 ) where {A,RT}
-    return NamedTuple{A,traced_type(RT, (), Val(mode))}((
+    return NamedTuple{A,traced_type(RT, (), Val(mode), track_numbers)}((
         (
             make_tracer(
-                seen, Base.getfield(prev, i), append_path(path, i), mode; kwargs...
+                seen,
+                Base.getfield(prev, i),
+                append_path(path, i),
+                mode;
+                track_numbers,
+                kwargs...,
             ) for i in 1:length(A)
         )...,
     ))

--- a/src/Tracing.jl
+++ b/src/Tracing.jl
@@ -12,7 +12,9 @@ for T in (DataType, Module, Nothing, Symbol, AbstractChar, AbstractString, RArra
     end
 end
 
-function traced_type(::Type{T}, seen, mode::Val{Mode}, track_numbers) where {T<:Number,Mode}
+function traced_type(
+    ::Type{T}, seen, mode::Val{Mode}, track_numbers
+) where {T<:Union{AbstractFloat,Integer},Mode}
     if Mode == ArrayToConcrete && any(Base.Fix1(<:, T), track_numbers)
         return ConcreteRNumber{T}
     end

--- a/test/autodiff.jl
+++ b/test/autodiff.jl
@@ -91,7 +91,7 @@ function cached_return(x, stret::StateReturn)
     return loss
 end
 
-function cached_return(x, stret::StateReturn2)
+function cached_return(x, stret::StateReturn1)
     loss = sum(x)
     tmp = x .+ 1
     stret.st1 = tmp
@@ -110,7 +110,7 @@ end
     @test stret.st isa ConcreteRArray
     @test stret.st ≈ x .+ 1
 
-    stret = StateReturn2(nothing, nothing)
+    stret = StateReturn1(nothing, nothing)
     ret = @jit Enzyme.gradient(Reverse, cached_return, x_ra, Const(stret))
 
     @test @allowscalar all(isone, ret[1])

--- a/test/autodiff.jl
+++ b/test/autodiff.jl
@@ -80,9 +80,22 @@ mutable struct StateReturn
     st::Any
 end
 
+mutable struct StateReturn1
+    st1::Any
+    st2::Any
+end
+
 function cached_return(x, stret::StateReturn)
     loss = sum(x)
     stret.st = x .+ 1
+    return loss
+end
+
+function cached_return(x, stret::StateReturn2)
+    loss = sum(x)
+    tmp = x .+ 1
+    stret.st1 = tmp
+    stret.st2 = tmp
     return loss
 end
 
@@ -96,4 +109,14 @@ end
     @test @allowscalar all(isone, ret[1])
     @test stret.st isa ConcreteRArray
     @test stret.st ≈ x .+ 1
+
+    stret = StateReturn2(nothing, nothing)
+    ret = @jit Enzyme.gradient(Reverse, cached_return, x_ra, Const(stret))
+
+    @test @allowscalar all(isone, ret[1])
+    @test stret.st1 isa ConcreteRArray
+    @test stret.st1 ≈ x .+ 1
+    @test stret.st2 isa ConcreteRArray
+    @test stret.st2 ≈ x .+ 1
+    @test stret.st1 === stret.st2
 end

--- a/test/autodiff.jl
+++ b/test/autodiff.jl
@@ -75,3 +75,25 @@ end
     @test typeof(res) == Tuple{Enzyme.TupleArray{ConcreteRNumber{Float64},(2, 2),4,2}}
     @test res[1] ≈ ones(2, 2)
 end
+
+mutable struct StateReturn
+    st::Any
+end
+
+function cached_return(x, stret::StateReturn)
+    loss = sum(x)
+    stret.st = x .+ 1
+    return loss
+end
+
+@testset "Cached Return: Issue #416" begin
+    x = rand(10)
+    x_ra = Reactant.to_rarray(x)
+
+    stret = StateReturn(nothing)
+    ret = @jit Enzyme.gradient(Reverse, cached_return, x_ra, Const(stret))
+
+    @test @allowscalar all(isone, ret[1])
+    @test stret.st isa ConcreteRArray
+    @test stret.st ≈ x .+ 1
+end

--- a/test/basic.jl
+++ b/test/basic.jl
@@ -676,3 +676,31 @@ end
     fill!(z, 1.0)
     @test all(==(1.0), Array(z))
 end
+
+@testset "Preserve Aliasing" begin
+    x = Reactant.to_rarray([3])
+    T = Any[nothing]
+
+    function ip(m, T)
+        @allowscalar m[1] = 2
+        T[1] = m
+        return m
+    end
+
+    res = @jit ip(x, T)
+    @test @allowscalar res[1] == 2
+    @test @allowscalar x[1] == 2
+    @test @allowscalar T[1][1] == 2
+
+    ptr_x = Base.unsafe_convert(
+        Ptr{Float64}, Reactant.XLA.UnsafeBufferPointer(x.data.buffer)
+    )
+    ptr_res = Base.unsafe_convert(
+        Ptr{Float64}, Reactant.XLA.UnsafeBufferPointer(res.data.buffer)
+    )
+    ptr_T1 = Base.unsafe_convert(
+        Ptr{Float64}, Reactant.XLA.UnsafeBufferPointer(T[1].data.buffer)
+    )
+
+    @test ptr_x == ptr_res == ptr_T1
+end

--- a/test/tracing.jl
+++ b/test/tracing.jl
@@ -90,7 +90,7 @@ using Test
                 (Val{:x}, Val{:x}),
             ]
                 tracedty = traced_type(
-                    origty, Reactant.OrderedIdDict(), Val(ConcreteToTraced)
+                    origty, Reactant.OrderedIdDict(), Val(ConcreteToTraced), ()
                 )
                 @test tracedty == targetty
             end
@@ -102,13 +102,13 @@ using Test
                 TracedRArray{Float64,3},
             ]
                 @test_throws Union{ErrorException,String} traced_type(
-                    type, Reactant.OrderedIdDict(), Val(ConcreteToTraced)
+                    type, Reactant.OrderedIdDict(), Val(ConcreteToTraced), ()
                 )
             end
         end
         @testset "traced_type exceptions" begin
             @test_throws TracedTypeError Reactant.traced_type(
-                Real, Reactant.OrderedIdDict(), Val(Reactant.ArrayToConcrete)
+                Real, Reactant.OrderedIdDict(), Val(Reactant.ArrayToConcrete), ()
             )
 
             struct Node
@@ -116,7 +116,7 @@ using Test
                 y::Union{Nothing,Node}
             end
             @test_throws NoFieldMatchError traced_type(
-                Node, Reactant.OrderedIdDict(), Val(ArrayToConcrete)
+                Node, Reactant.OrderedIdDict(), Val(ArrayToConcrete), ()
             )
         end
     end

